### PR TITLE
WIP: Define schema with cue

### PR DESCRIPTION
**Pull request by embano1**
TODO:

- [ ] Discuss folder structure
- [ ] Documentation, examples and descriptions
- [ ] Add `indicator`
- [ ] Add `objectives`

Closes: #20
Signed-off-by: Michael Gasch <mgasch@vmware.com>

### CODE REVIEW

**Comment by embano1** [comment]
@ian-bartholomew first proposal for discussion. There's couple of areas where the spec is a bit fuzzy/unclear to me (and I made some minor adjustments based on some other comments).

On `indicator` and `objectives` I had a couple of questions thus did not include them in the spec yet.

> indicator optional, represents the Service Level Indicator (SLI). Currently this only supports one Metric, thresholdMetric, with ratioMetric supported in the objectives stanza.

Why do we need to spread the different metrics over various spec parts, e.g. one in `indicator`, one in `objectives`?
This also leads to some tough validation requirements e.g.

> If thresholdMetric has been defined, only one Threshold can be defined. However if using ratioMetric then any number of Thresholds can be defined.

and

> indicator.ratioMetric Metric {Good, Total}, if ratioMetric is defined then thresholdMetric should not be set in indicator

File: spec/openapi.json [hunk 1]
```diff
@@ -0,0 +1,198 @@
+{
```
Comment (embano1) [inline spec/openapi.json:1]: Generated with `cue export --out=openapi meta.cue spec.cue > openapi.json`

File: spec/slo.cue [hunk 2]
```diff
@@ -0,0 +1,6 @@
+package v1alpha1
+
+apiVersion: "openslo/v1alpha1" & #apiVersion
```
Comment (embano1) [inline spec/slo.cue:3]: enforcing `v1alpha1` semantics

File: spec/slo.cue [hunk 3]
```diff
@@ -0,0 +1,6 @@
...
+kind:       "SLO" & #kind
```
Comment (embano1) [inline spec/slo.cue:4]: For `Service` the spec is rather brief on what goes into it (besides `spec.Description`) so I did not include `Service` type yet.

File: spec/spec.cue [hunk 4]
```diff
@@ -0,0 +1,50 @@
+package v1alpha1
+
+import "time"
+
+#spec: {
+	// list?
```
Comment (embano1) [inline spec/spec.cue:6]: In the spec it looks like a list `[]`?

File: spec/spec.cue [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+	service: string
+	// optional
+	// TODO: 1050 exceeds go regex limit (1000)
+	description?: string
```
Comment (embano1) [inline spec/spec.cue:10]: Can't use regex to validate 1050 char limit here due to go regex package limit (1000).

Comment (embano1) [inline spec/spec.cue:10]: Can't use regex to validate 1050 char limit here due to go regex package limit (1000).

File: spec/spec.cue [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+	// object oneOf
+	timeWindow:      #timeWindowSecond | #timeWindowCalendar
```
Comment (embano1) [inline spec/spec.cue:12]: Note: made this oneOf instead of list

File: spec/spec.cue [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	budgetingMethod: "Occurrences" | "Timeslices"
+    // objectives:
+    // indicator:
+}
+
+#indicator: {
```
Comment (embano1) [inline spec/spec.cue:18]: See questions in PR comments.

File: spec/spec.cue [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+}
+
+#timeWindowCalendar: {
+	#timeWindow & {
+		unit:  "Year" | "Quarter" | "Month" | "Week" | "Day"
+		count: "numeric"
+	}
+	calendar: {
+		// note: leap-seconds not supported
+		startTime: time.Format(time.RFC3339)
```
Comment (embano1) [inline spec/spec.cue:48]: Enforcing RFC3339 and avoiding additional timezone field.

**Comment by its02003** [comment]
> @ian-bartholomew first proposal for discussion. There's couple of areas where the spec is a bit fuzzy/unclear to me (and I made some minor adjustments based on some other comments).
> 
> On `indicator` and `objectives` I had a couple of questions thus did not include them in the spec yet.
> 
> > indicator optional, represents the Service Level Indicator (SLI). Currently this only supports one Metric, thresholdMetric, with ratioMetric supported in the objectives stanza.
> 
> Why do we need to spread the different metrics over various spec parts, e.g. one in `indicator`, one in `objectives`?
> This also leads to some tough validation requirements e.g.
> 
> > If thresholdMetric has been defined, only one Threshold can be defined. However if using ratioMetric then any number of Thresholds can be defined.
> 
> and
> 
> > indicator.ratioMetric Metric {Good, Total}, if ratioMetric is defined then thresholdMetric should not be set in indicator

Agreed as I mentioned this here: https://github.com/OpenSLO/OpenSLO/issues/29

Whats the best mechanism for suggesting the recommended change to the spec?

**Comment by ian-bartholomew** [comment]
> Why do we need to spread the different metrics over various spec parts, e.g. one in indicator, one in objectives?

@embano1 this has come up a few times, and I agree, it leads to some issues, and should get addressed.  This is an artifact of us open sourcing our internal spec

@its02003 I think a good way of doing is it is a PR here with changes in the spec, and then pinging the OpenSLO Slack channel where we can discuss it.  

Thanks both!

**Pull request closed without merging**